### DISCUSSION TITLE
Build DPC++ toolchain with libc++ library undefined symbol `_ZN2cl4sycl3ext6oneapi10level_zero12make_contextERKSt6vectorINS0_6deviceESaIS5_EEmb' referenced in expression

### DISCUSSION THREAD

**Discussion by haonanya** [intel/llvm#4926] [Q&A] [answered]
Build DPC++ toolchain with libc++ library, some link error occured: 
`[9/41] Linking CXX shared library lib/libsycl.so.5.4.0-0
FAILED: lib/libsycl.so.5.4.0-0
: && /usr/bin/c++ -fPIC -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-comment -Wmisleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wextra -Wno-deprecated-declarations -Werror -O3 -DNDEBUG  -Wl,-z,defs -Wl,-z,nodelete -shared -Wl,-soname,libsycl.so.5 -o lib/libsycl.so.5.4.0-0 tools/sycl/source/CMakeFiles/sycl_object.dir/backend/opencl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/backend/level_zero.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/backend.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/accessor_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/allowlist.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/buffer_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/builtins_common.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/builtins_geometric.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/builtins_integer.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/builtins_math.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/builtins_relational.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/pi.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/common.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/config.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/context_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/device_binary_image.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/device_filter.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/device_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/error_handling/enqueue_kernel.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/event_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/filter_selector_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/force_device.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/global_handler.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/helpers.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/handler_proxy.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/image_accessor_util.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/image_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/kernel_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/kernel_program_cache.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/memory_manager.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/platform_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/program_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/program_manager/program_manager.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/queue_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/online_compiler/online_compiler.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/os_util.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/persistent_device_code_cache.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/platform_util.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/reduction.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/sampler_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/stream_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/scheduler/commands.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/scheduler/leaves_collection.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/scheduler/scheduler.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/scheduler/scheduler_helpers.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/scheduler/graph_processor.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/scheduler/graph_builder.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/spec_constant_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/sycl_mem_obj_t.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/usm/usm_impl.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/util.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/accessor.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/context.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/device.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/device_selector.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/event.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/exception.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/exception_list.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/function_pointer.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/half_type.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/handler.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/interop_handle.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/interop_handler.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/kernel.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/kernel_bundle.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/platform.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/program.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/queue.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/sampler.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/stream.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/spirv_ops.cpp.o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/posix_pi.cpp.o  -Wl,-rpath,/home/haonanya/workspace/llvm/buildWithLibcpp/lib:  lib/libxpti.a  -Wl,/home/haonanya/workspace/llvm/sycl/source/abi_replacements_linux.txt  -Wl,--version-script=/home/haonanya/workspace/llvm/sycl/source/ld-version-script.txt  -ldl  lib/libOpenCL.so.1.2  -ldl  -lpthread  -L/home/haonanya/workspace/llvm-project/build/lib/  -Wl,-rpath,/home/haonanya/workspace/llvm-project/build/lib/  -nodefaultlibs  -lc++  -lc++abi  -lm  -lc  -lgcc_s  -lgcc && :
/home/haonanya/workspace/llvm/sycl/source/abi_replacements_linux.txt:4: undefined symbol `_ZN2cl4sycl3ext6oneapi10level_zero12make_contextERKSt6vectorINS0_6deviceESaIS5_EEmb' referenced in expression
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.`

I have build libc++ with LLVM 12, and I follow the steps on https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md#build-dpc-toolchain-with-libc-library. I remember that it was build successful long time ago. When I pull latest changes, it begins to build failed, could you please help to figure out how to resolve the link error?
Thanks very much.

**Comment by alexbatashev** [comment]
@haonanya I suggest you commenting these lines as a workaround: https://github.com/intel/llvm/blob/0c33048e492661e50596eb5f8650eeff92347db9/sycl/source/CMakeLists.txt#L55-L57

**Reply by haonanya** [reply]
Hi, @alexbatashev. Thanks for your comments, but it has failures on buildbot config stages.
[CMakeError.log](https://github.com/intel/llvm/files/7510603/CMakeError.log)

**Reply by alexbatashev** [reply]
There's something wrong with your local configuration. CMake tries to compile a simple program before it actually generates build files, and this is what failed in your case.

**Reply by haonanya** [reply]
Hi, @alexbatashev, I am not familiar with CMake, could you please help to figure out what is the right approach? I'd appreciate it if you have some comments.
I updated CMakeList.txt, then run config:
`python3 ../buildbot/configure.py --use-libcxx --libcxx-include ~/workspace/llvm-project/build/include/c++/v1 --libcxx-library ~/workspace/llvm-project/build/lib/ -o ~/workspace/llvm/buildWithLibcpp/`

I checked that libcxx include path and library path is right, but I got the errors as attached CmakeError.log.
Thanks very much.

**Reply by linuxcpp6** [reply]
Is it solved?
python3 ../buildbot/configure.py --use-libcxx --libcxx-include ~/workspace/llvm-project/build/include/c++/v1 --libcxx-library ~/workspace/llvm-project/build/lib/ -o ~/workspace/llvm/buildWithLibcpp/
Is it right or not?

**Reply by linuxcpp6** [reply]
I hope to get your reply

**Reply by haonanya** [reply]
Yes, command is right, please update CMakeList.txt using the workaround.

**Reply by linuxcpp6** [reply]
This CMakeList.txt alrealy has these four lines.

**Reply by alexbatashev** [reply]
@TechnologyAndEducation you should comment them out.

**Reply by linuxcpp6** [reply]
> @TechnologyAndEducation you should comment them out.
Thanks

**Reply by linuxcpp6** [reply]
> it solved

**Comment by haonanya** [comment, marked as answer]
Thanks @alexbatashev offer workaround.
https://github.com/intel/llvm/blob/0c33048e492661e50596eb5f8650eeff92347db9/sycl/source/CMakeLists.txt#L55-L58
